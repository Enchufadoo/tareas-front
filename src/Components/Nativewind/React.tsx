import { styled as nativewindStyled } from 'nativewind'
import * as ReactNative from 'react-native'
import * as VectorIcons from '@expo/vector-icons'
import { styled } from 'dripsy'

export const Text = styled(nativewindStyled(ReactNative.Text))({
  color: '$text'
})

export const View = styled(nativewindStyled(ReactNative.View))()

export const Pressable = styled(nativewindStyled(ReactNative.Pressable))()
export const TextInput = styled(nativewindStyled(ReactNative.TextInput))()
export const TouchableOpacity = styled(
  nativewindStyled(ReactNative.TouchableOpacity)
)()
export const Switch = styled(nativewindStyled(ReactNative.Switch))()
export const SimpleLineIcons = styled(
  nativewindStyled(VectorIcons.SimpleLineIcons)
)()
export const MaterialIcons = styled(
  nativewindStyled(VectorIcons.MaterialIcons)
)()

export const Entypo = styled(nativewindStyled(VectorIcons.Entypo))()

export const Octicons = styled(nativewindStyled(VectorIcons.Octicons))()
