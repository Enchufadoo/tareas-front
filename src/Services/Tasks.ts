import { api } from './Api'
import moment from 'moment'

export type Task = {
  id: number
  title: string
  description: string
  created_at: string
  created_at_date: string
  hours_since_progress: number
  progress_today: boolean
  last_progress?: {
    created_at: string
    id: number
  }
}

export type TaskCount = { finished: number; unfinished: number }

export enum TaskProgressState {
  open,
  finished,
  all
}

interface TaskWithUser extends Task {
  user: {
    username: string
    email: string
    avatar: string
  }
}

export type GetTasksResponse = {
  data: {
    tasks: Task[]
    task_count: TaskCount
  }
}

type GetTaskResponse = {
  data: {
    task: TaskWithUser
  }
}

type getCreateTaskDataResponse = {
  data: {
    friends: string[]
  }
}
type setAddTaskProgressResponse = {
  data: {}
}
type setRemoveTaskProgressResponse = {
  data: {}
}

type setDeleteTaskResponse = {
  data: {}
}

type CreateTaskData = {
  title: string
  description: string
}

export enum FinishTask {
  unfinished,
  finished
}

type FinishTaskData = {
  taskId: number
  finishTask: FinishTask
}

type setCreateTaskResponse = {}

export const tasksApi = api.injectEndpoints({
  overrideExisting: true,
  endpoints: (build) => ({
    getTasks: build.query<GetTasksResponse, number>({
      providesTags: ['TasksList'],
      query: (finished: number) => {
        return {
          url: `/task`,
          method: 'GET',
          params: {
            finished: finished
          }
        }
      }
    }),
    getTask: build.query<GetTaskResponse, number>({
      query: (id: number) => {
        return {
          url: `/task/${id}`,
          method: 'GET'
        }
      },
      transformResponse: (response: GetTaskResponse) => {
        response.data.task.created_at_date = moment().format('YYYY-MM-DD')
        return response
      },
      providesTags: ['TaskView']
    }),
    setCreateTask: build.mutation<setCreateTaskResponse, CreateTaskData>({
      query: (taskData) => {
        return {
          url: `/task`,
          method: 'POST',
          data: taskData
        }
      },
      invalidatesTags: ['TasksList']
    }),
    setAddTaskProgress: build.mutation<setAddTaskProgressResponse, number>({
      query: (taskId) => {
        return {
          url: `/task/progress/${taskId}`,
          method: 'POST',
          data: {}
        }
      },
      invalidatesTags: ['TaskView', 'TasksList']
    }),
    setRemoveTaskProgress: build.mutation<
      setRemoveTaskProgressResponse,
      number
    >({
      query: (taskId) => {
        return {
          url: `/task/progress/${taskId}`,
          method: 'DELETE',
          data: {}
        }
      },
      invalidatesTags: ['TaskView', 'TasksList']
    }),
    setDeleteTask: build.mutation<setDeleteTaskResponse, number>({
      query: (taskId) => {
        return {
          url: `/task/${taskId}`,
          method: 'DELETE',
          data: {}
        }
      },
      invalidatesTags: ['TasksList']
    }),
    setFinishTask: build.mutation<setDeleteTaskResponse, FinishTaskData>({
      query: (finishTaskData) => {
        return {
          url: `/task/finish/${finishTaskData.taskId}/${finishTaskData.finishTask}`,
          method: 'PUT',
          data: {}
        }
      },
      invalidatesTags: ['TasksList']
    }),
    getCreteTaskData: build.query<getCreateTaskDataResponse, void>({
      query: () => {
        return {
          url: `/task/store/data`,
          method: 'GET'
        }
      }
    }),
    getWeeksTasks: build.query<
      {
        data: {
          progress: {
            [key: number]: boolean[]
          }
          tasks: {
            id: number
            title: string
          }[]
        }
      },
      void
    >({
      query: () => {
        return {
          url: `/task/week`,
          method: 'GET'
        }
      },
      providesTags: []
    })
  })
})

export const {
  useGetTasksQuery,
  useGetTaskQuery,
  useGetCreteTaskDataQuery,
  useSetCreateTaskMutation,
  useSetAddTaskProgressMutation,
  useSetRemoveTaskProgressMutation,
  useSetFinishTaskMutation,
  useLazyGetWeeksTasksQuery
} = tasksApi
