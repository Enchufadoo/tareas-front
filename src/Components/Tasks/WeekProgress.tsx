import React from 'react'

import { Entypo, Text, View } from '@/Components/Nativewind/React'
import { StyleProp } from 'react-native/Libraries/StyleSheet/StyleSheet'
import { ViewStyle } from 'react-native/Libraries/StyleSheet/StyleSheetTypes'
import { useSx } from 'dripsy'

const BORDER_WIDTH = 0.3

type Props = {
  progress: boolean[]
  isLastElement: boolean
}

const WeekProgress = (props: Props) => {
  const sx = useSx()

  let rowStyle: StyleProp<ViewStyle> = {
    flexDirection: 'row',
    borderTopWidth: BORDER_WIDTH
  }

  if (props.isLastElement) {
    rowStyle.borderBottomWidth = BORDER_WIDTH
  }

  return (
    <View style={rowStyle}>
      {props.progress.map((day, index) => {
        let style: StyleProp<ViewStyle> = {
          borderRightWidth: BORDER_WIDTH,
          borderColor: 'black'
        }

        if (index === 0) {
          style.borderLeftWidth = BORDER_WIDTH
        }
        return (
          <View
            key={index}
            className={'justify-center items-center p-2 flex-1'}
            style={style}
          >
            {day ? (
              <Entypo
                name={'cross'}
                size={24}
                color={sx({ color: '$success' })['color']}
              />
            ) : (
              <Text>{''}</Text>
            )}
          </View>
        )
      })}
    </View>
  )
}

export default WeekProgress
