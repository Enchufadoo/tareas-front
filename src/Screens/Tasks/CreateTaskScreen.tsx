import { Text, View } from '@/Components/Nativewind/React'
import {
  useGetCreteTaskDataQuery,
  useSetCreateTaskMutation
} from '@/Services/Tasks'
import { setLoadingData } from '@/Store/Features/ApplicationSlice'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import React, { useEffect } from 'react'
import { ScrollView } from 'react-native'
import { useDispatch } from 'react-redux'
import { Controller, useForm } from 'react-hook-form'
import { isValidForm } from '@/Util/FormUtil'
import InputError from '@/Components/Errors/InputError'

import Button, {
  ButtonSize,
  ButtonTheme
} from '@/Components/Base/Button/Button'
import FormSubmitError from '@/Components/Errors/FormSubmitError'
import { MultilineInput } from '@/Components/Base/Input/MultilineInput'
import { TextInput } from '@/Components/Base/Input/TextInput'
import BaseLayout from '@/Components/Layout/BaseLayout/BaseLayout'
import { LoggedInStackParamList } from '@/Router/LoggedInStackNavigator'

type Props = NativeStackScreenProps<LoggedInStackParamList, 'Create Task'>

const CreateTaskScreen = (props: Props) => {
  const { isFetching, isSuccess, isError, data } = useGetCreteTaskDataQuery()

  const [setCreateTask, setCreateTaskResult] = useSetCreateTaskMutation()

  const dispatch = useDispatch()

  const requiredFields = ['title']

  const {
    getValues,
    control,
    formState: { errors, isDirty, dirtyFields }
  } = useForm({
    mode: 'onChange',
    defaultValues: {
      title: '',
      description: ''
    }
  })

  let isValid = isValidForm(errors, requiredFields, dirtyFields)

  useEffect(() => {
    if (setCreateTaskResult.isSuccess) {
      props.navigation.navigate('Home Screen')
    }
  }, [setCreateTaskResult.isSuccess])

  useEffect(() => {
    dispatch(setLoadingData(isFetching))
  }, [isFetching])

  return (
    <>
      {isSuccess && (
        <ScrollView style={{ flex: 1 }} contentContainerStyle={{ flexGrow: 1 }}>
          <BaseLayout>
            <View className="flex-1">
              <View className="mb-5">
                <Text className="text-2xl font-bold">New Task</Text>
                <Text>Create a new task</Text>
              </View>
              <View className="flex-1 space-y-4">
                <View>
                  <Controller
                    control={control}
                    rules={{
                      required: {
                        value: true,
                        message: 'The title field is required'
                      },
                      minLength: {
                        value: 4,
                        message: 'The title must be at least 4 characters long'
                      },
                      maxLength: {
                        value: 30,
                        message: 'The title must be at most 30 characters'
                      }
                    }}
                    render={({ field: { onChange, onBlur, value } }) => (
                      <TextInput
                        placeholder={'Title'}
                        onBlurForm={onBlur}
                        onChangeText={onChange}
                        value={value}
                        keyboardType="default"
                        isValid={control.getFieldState('title')}
                      />
                    )}
                    name="title"
                  />

                  <InputError error={errors.title} />
                </View>
                <View>
                  <Controller
                    control={control}
                    rules={{
                      required: {
                        value: false,
                        message: 'The description field is required'
                      },
                      minLength: {
                        value: 4,
                        message:
                          'The description must be at least 4 characters long'
                      },
                      maxLength: {
                        value: 3000,
                        message: 'The title must be at most 3000 characters'
                      }
                    }}
                    render={({ field: { onChange, onBlur, value } }) => (
                      <MultilineInput
                        height={300}
                        placeholder={'Description'}
                        onBlurForm={onBlur}
                        onChangeText={onChange}
                        value={value}
                        keyboardType="default"
                        isValid={control.getFieldState('description')}
                      />
                    )}
                    name="description"
                  />

                  <InputError error={errors.title} />
                </View>
                {setCreateTaskResult.isError && (
                  <View className={'mt-6'}>
                    <FormSubmitError
                      header={'Creating a task failed'}
                      subheader={'Please try again later'}
                    />
                  </View>
                )}
              </View>
            </View>
            <View className={'mt-12 mb-12'}>
              <Button
                theme={ButtonTheme.primary}
                size={ButtonSize.medium}
                title={'Create Task'}
                disabled={!isDirty || !isValid}
                outline={false}
                loading={
                  setCreateTaskResult.isLoading || setCreateTaskResult.isSuccess
                }
                onPress={async () => {
                  await setCreateTask(getValues())
                }}
              />
            </View>
          </BaseLayout>
        </ScrollView>
      )}
    </>
  )
}

export default CreateTaskScreen
