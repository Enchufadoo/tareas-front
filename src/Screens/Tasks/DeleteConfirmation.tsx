import React from 'react'
import { View } from 'react-native'

type DialogComponentProps = {
  visible: boolean
  toggleVisible: () => void
  deleteAction: () => void
}

const DeleteConfirmation: React.FunctionComponent<DialogComponentProps> = (
  props
) => {
  return (
    <View>
      {/*<Dialog isVisible={props.visible} onBackdropPress={props.toggleVisible}>
        <Dialog.Title title="Delete Task Confirmation" />
        <Text>Are you sure you want to delete this task</Text>
        <Dialog.Actions>
          <Dialog.Button title="Delete" onPress={props.deleteAction} />
          <Dialog.Button title="Cancel" onPress={() => props.toggleVisible()} />
        </Dialog.Actions>
      </Dialog>*/}
    </View>
  )
}

export default DeleteConfirmation
