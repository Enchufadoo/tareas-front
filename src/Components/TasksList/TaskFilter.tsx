import { TaskCount, TaskProgressState } from '@/Services/Tasks'
import { jc } from '@/Util/NativewindUtil'
import React, { useEffect, useState } from 'react'
import { Text, TouchableOpacity, View } from '../Nativewind/React'
import Badge from './Badge'

interface CountResult {
  number: number
}

type Props = {
  taskFilterState: number
  setTaskFilterState: (arg0: number) => void
  tasksCount?: TaskCount
}

export const listStates = [
  { value: TaskProgressState.open, text: 'Open' },
  { value: TaskProgressState.finished, text: 'Finished' },
  { value: TaskProgressState.all, text: 'All' }
]

const calculateCount = (taskCount: TaskCount) => {
  let result = {} as CountResult

  let finished = taskCount['finished']
  let unfinished = taskCount['unfinished']

  result[TaskProgressState.finished] = finished
  result[TaskProgressState.open] = unfinished
  result[TaskProgressState.all] = finished + unfinished

  return result
}

const TaskFilter = (props: Props) => {
  let textClassList = ['text-lg ']

  const [count, setCount] = useState<CountResult>()

  useEffect(() => {
    if (props.tasksCount) {
      setCount(calculateCount(props.tasksCount))
    }
  }, [props.tasksCount])

  const renderStates = listStates.map((el, index) => {
    let textClassListCopy = textClassList.slice()
    let selected = props.taskFilterState === el.value

    return (
      <View key={el.text} className="flex-row flex-1 py-3">
        <TouchableOpacity
          key={el.text}
          className="focus:outline-none flex-row flex-1 justify-center items-center"
          onPress={() => {
            props.setTaskFilterState(el.value)
          }}
        >
          <View className={'flex-row flex-1  items-center'}>
            {count && (
              <Badge
                count={count[el.value]}
                selected={selected}
                style={{ marginRight: 6 }}
              />
            )}
            <Text
              sx={{
                color: selected ? '$primary' : '$text'
              }}
              className={jc(textClassListCopy)}
            >
              {el.text}
            </Text>
          </View>
        </TouchableOpacity>
      </View>
    )
  })

  return (
    <View className="flex-row justify-between mt-3 mb-3 text-center mr-2">
      {renderStates}
    </View>
  )
}

export default TaskFilter
