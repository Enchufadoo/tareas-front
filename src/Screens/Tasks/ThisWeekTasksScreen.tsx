import React, { useCallback } from 'react'
import BaseLayout from '@/Components/Layout/BaseLayout/BaseLayout'
import { Text, View } from '@/Components/Nativewind/React'
import { useLazyGetWeeksTasksQuery } from '@/Services/Tasks'
import { useLoadingData } from '@/Lib/UseLoadingData'
import { useFocusEffect } from '@react-navigation/native'
import WeekProgress from '@/Components/Tasks/WeekProgress'

const ThisWeekTasksScreen = () => {
  const [getWeeksTasksTrigger, getWeeksTasksResult] =
    useLazyGetWeeksTasksQuery()

  useFocusEffect(
    useCallback(() => {
      getWeeksTasksTrigger()
    }, [])
  )

  useLoadingData([getWeeksTasksResult])

  return (
    <BaseLayout>
      <Text className="text-2xl font-bold">Progress This Week</Text>

      {getWeeksTasksResult.isSuccess &&
        !getWeeksTasksResult.isFetching &&
        (getWeeksTasksResult.data?.data?.tasks?.length ? (
          <View className={'flex-1 my-2'}>
            <View className="flex-row">
              <View style={{ flex: 1 }}></View>
              <View style={{ flex: 3 }}>
                <View className={'flex-row'}>
                  {['S', 'M', 'T', 'W', 'T', 'F', 'S'].map((day, index) => {
                    return (
                      <View className={'flex-1 items-center'} key={index}>
                        <Text>{day}</Text>
                      </View>
                    )
                  })}
                </View>
              </View>
            </View>
            {getWeeksTasksResult.data.data.tasks.map((task, index) => {
              return (
                <View className={'flex-row'} key={task.id}>
                  <View className={'flex-1 justify-center '}>
                    <Text numberOfLines={2}>{task.title}</Text>
                  </View>
                  <View style={{ flex: 3 }}>
                    <WeekProgress
                      progress={getWeeksTasksResult.data.data.progress[task.id]}
                      isLastElement={
                        index === getWeeksTasksResult.data.data.tasks.length - 1
                      }
                    />
                  </View>
                </View>
              )
            })}
          </View>
        ) : (
          <View className="items-center flex-1 justify-center basis-4/5 shrink-0 ">
            <View className="flex-row justify-center">
              <Text
                className="text-2xl font-medium "
                sx={{
                  color: '$lightText'
                }}
              >
                You haven't created any task
              </Text>
            </View>
          </View>
        ))}
    </BaseLayout>
  )
}

export default ThisWeekTasksScreen
