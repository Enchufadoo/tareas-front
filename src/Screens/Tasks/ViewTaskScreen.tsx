import { Entypo, Text, View } from '@/Components/Nativewind/React'

import { FinishTask, tasksApi } from '@/Services/Tasks'
import { capitalize } from '@/Util/StringUtil'
import { useActionSheet } from '@expo/react-native-action-sheet'
import { NativeStackScreenProps } from '@react-navigation/native-stack'

import React, { useEffect, useState } from 'react'
import { TouchableOpacity } from 'react-native-gesture-handler'
import DeleteConfirmation from './DeleteConfirmation'
import Button, {
  ButtonSize,
  ButtonTheme
} from '@/Components/Base/Button/Button'
import BaseLayout from '@/Components/Layout/BaseLayout/BaseLayout'
import { withErrorBoundary } from '@/Components/Errors/RoutesErrorBoundary'
import { LoggedInStackParamList } from '@/Router/LoggedInStackNavigator'
import { useLoadingData } from '@/Lib/UseLoadingData'
import { useSx } from 'dripsy'

type Props = NativeStackScreenProps<LoggedInStackParamList, 'View Task'>
const ViewTaskScreen = (props: Props) => {
  const getTasksResult = tasksApi.useGetTaskQuery(props.route.params.id)
  const sx = useSx()

  const { isSuccess, data } = getTasksResult

  const { showActionSheetWithOptions } = useActionSheet()

  const [setAddTaskProgress, setAddTaskProgressResult] =
    tasksApi.useSetAddTaskProgressMutation()

  const [setDeleteTaskMutation, setDeleteTaskResult] =
    tasksApi.useSetDeleteTaskMutation()

  const [setFinishTaskMutation, setFinishTaskResult] =
    tasksApi.useSetFinishTaskMutation()

  const task = data?.data.task

  const onSetProgress = async () => {
    await setAddTaskProgress(task.id)
  }

  const [deleteDialogVisible, setDeleteDialogVisible] = useState(false)

  useLoadingData([
    getTasksResult,
    setAddTaskProgressResult,
    setDeleteTaskResult
  ])

  useEffect(() => {
    if (isSuccess) {
      props.navigation.setOptions({
        title: 'Task details',
        headerRight: () => {
          return (
            <View className="px-8">
              <TouchableOpacity onPress={showTasksMenu}>
                <Entypo
                  sx={{ color: '$text' }}
                  name="dots-three-horizontal"
                  size={24}
                  color="black"
                />
              </TouchableOpacity>
            </View>
          )
        }
      })
    }
  }, [isSuccess])

  const showTasksMenu = () => {
    const options = ['Delete Task', 'Finish Task', 'Cancel']

    const destructiveButtonIndex = 0
    const finishTaskIndex = 1
    const cancelButtonIndex = 2

    showActionSheetWithOptions(
      {
        options,
        cancelButtonIndex,
        destructiveButtonIndex,
        containerStyle: sx({ backgroundColor: '$background' }),
        textStyle: sx({ color: '$text' }),
        destructiveColor: sx({ color: '$text' })['color']
      },
      (selectedIndex: number) => {
        switch (selectedIndex) {
          case destructiveButtonIndex:
            setDeleteTaskMutation(task.id)
            props.navigation.navigate('Home Screen')

            setDeleteDialogVisible(true)
            break
          case finishTaskIndex:
            setFinishTaskMutation({
              finishTask: FinishTask.finished,
              taskId: props.route.params.id
            })
            break
          case cancelButtonIndex:
          // Canceled
        }
      }
    )
  }

  return (
    <BaseLayout>
      {isSuccess && (
        <View className="flex-1 justify-between">
          <View className=" flex-1">
            <View className="flex-row"></View>
            <View className="mt-5">
              <Text className="text-xl">
                {capitalize(data.data.task.title)}
              </Text>
            </View>
            <View className="mt-5">
              {data.data.task.description ? (
                <>
                  <Text className="font-semibold text-sm">Description</Text>
                  <Text className="text-md">{data.data.task.description}</Text>
                </>
              ) : (
                <>
                  <Text className="font-semibold text-sm">No description</Text>
                </>
              )}
            </View>
            <DeleteConfirmation
              deleteAction={async () => {
                await setDeleteTaskMutation(task.id)
                props.navigation.navigate('Home Screen')
              }}
              visible={deleteDialogVisible}
              toggleVisible={() => {
                setDeleteDialogVisible(!deleteDialogVisible)
              }}
            ></DeleteConfirmation>

            <View className="mt-7">
              {data.data.task.progress_today ? (
                <View
                  className="flex flex-row w-full max-w-lg overflow-hidden  border border-solid"
                  sx={{ borderColor: '$lightText' }}
                >
                  <View
                    className="w-1"
                    sx={{ backgroundColor: '$success' }}
                  ></View>

                  <View className="flex px-2 py-3 w-full flex-row justify-between">
                    <View>
                      <Text className="font-bold mx-3 mb-1">Status</Text>
                      <Text className=" mx-3">Progress added today</Text>
                    </View>
                  </View>
                </View>
              ) : (
                <View
                  className="flex flex-row w-full max-w-lg overflow-hidden border border-solid"
                  sx={{ borderColor: '$lightText' }}
                >
                  <View
                    className="w-1"
                    sx={{ backgroundColor: '$text' }}
                  ></View>
                  <View className="flex px-2 py-3 w-full">
                    <Text
                      className="font-bold mx-3 mb-1"
                      sx={{ color: '$text' }}
                    >
                      Status
                    </Text>
                    <Text className="mx-3" sx={{ color: '$text' }}>
                      Task still pending
                    </Text>
                  </View>
                </View>
              )}
            </View>
          </View>
          <View>
            <Button
              size={ButtonSize.medium}
              theme={ButtonTheme.primary}
              onPress={() => {
                onSetProgress()
              }}
              title="Add Progress"
            />
          </View>
        </View>
      )}
    </BaseLayout>
  )
}

export default withErrorBoundary(ViewTaskScreen)
