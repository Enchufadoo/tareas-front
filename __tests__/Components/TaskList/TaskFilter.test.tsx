import React from 'react'
import TaskFilter, { listStates } from '@/Components/TasksList/TaskFilter'
import { NavigationCTC, renderWithProviders } from '@/Testing/TestUtil'
import { TaskProgressState } from '@/Services/Tasks'
import { fireEvent } from '@testing-library/react-native'
import renderer from 'react-test-renderer'

describe('TaskFilter', () => {
  it('renders without crashing', () => {
    const { getByText, toJSON } = renderWithProviders(
      <TaskFilter
        tasksCount={{ finished: 5, unfinished: 10 }}
        taskFilterState={0}
        setTaskFilterState={() => {}}
      />
    )

    expect(getByText('10')).toBeTruthy()
    expect(getByText('5')).toBeTruthy()
    expect(getByText('15')).toBeTruthy()
  })

  it('sets the filtered state on press', () => {
    const setTaskFilterState = jest.fn()

    const { getByText } = renderWithProviders(
      <TaskFilter
        taskFilterState={TaskProgressState.open}
        setTaskFilterState={setTaskFilterState}
        tasksCount={{ finished: 8, unfinished: 0 }}
      />
    )

    listStates.forEach((state, index) => {
      fireEvent.press(getByText(state.text))

      expect(setTaskFilterState).toHaveBeenCalledWith(index)
    })
  })

  it('the filtered state doesnt change when same state is clicked', () => {
    const setTaskFilterState = jest.fn()

    const { getByText } = renderWithProviders(
      <TaskFilter
        taskFilterState={TaskProgressState.finished}
        setTaskFilterState={setTaskFilterState}
        tasksCount={{ finished: 8, unfinished: 0 }}
      />
    )

    fireEvent.press(getByText(listStates[TaskProgressState.finished].text))

    expect(setTaskFilterState).toHaveBeenCalledWith(
      listStates[TaskProgressState.finished].value
    )
  })

  it('Snapshot test', () => {
    const setTaskFilterState = jest.fn()
    const component = renderer
      .create(
        <NavigationCTC>
          <TaskFilter
            taskFilterState={TaskProgressState.open}
            setTaskFilterState={setTaskFilterState}
            tasksCount={{ finished: 8, unfinished: 0 }}
          />
        </NavigationCTC>
      )
      .toJSON()

    expect(component).toMatchSnapshot()
  })
})
