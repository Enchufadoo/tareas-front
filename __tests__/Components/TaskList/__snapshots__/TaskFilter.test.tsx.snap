// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskFilter Snapshot test 1`] = `
<View
  style={
    [
      [
        {},
        undefined,
        {},
      ],
    ]
  }
>
  <View
    style={
      [
        [
          {},
          undefined,
          {},
        ],
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            [
              {},
              undefined,
              {},
            ],
          ]
        }
      >
        <Text
          style={
            [
              [
                {
                  "color": "#000",
                },
                undefined,
                {
                  "color": "#b336be",
                },
              ],
            ]
          }
        >
          Open
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      [
        [
          {},
          undefined,
          {},
        ],
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            [
              {},
              undefined,
              {},
            ],
          ]
        }
      >
        <Text
          style={
            [
              [
                {
                  "color": "#000",
                },
                undefined,
                {
                  "color": "#000",
                },
              ],
            ]
          }
        >
          Finished
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      [
        [
          {},
          undefined,
          {},
        ],
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            [
              {},
              undefined,
              {},
            ],
          ]
        }
      >
        <Text
          style={
            [
              [
                {
                  "color": "#000",
                },
                undefined,
                {
                  "color": "#000",
                },
              ],
            ]
          }
        >
          All
        </Text>
      </View>
    </View>
  </View>
</View>
`;
